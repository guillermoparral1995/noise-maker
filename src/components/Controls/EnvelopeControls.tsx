import React from 'react';
import Knob from '../shared/Knob';
import {
  updateAttack,
  updateDecay,
  updateRelease,
  updateSustain,
} from '../../store/actions';

const EnvelopeFilterControls = () => {
  return (
    <div>
      <Knob
        label="attack"
        defaultValue={0}
        action={updateAttack}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="decay"
        defaultValue={0}
        action={updateDecay}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="sustain"
        defaultValue={1}
        action={updateSustain}
        min={0}
        max={1}
      ></Knob>
      <Knob
        label="release"
        defaultValue={0}
        action={updateRelease}
        min={0}
        max={5}
      ></Knob>
    </div>
  );
};

export default EnvelopeFilterControls;
