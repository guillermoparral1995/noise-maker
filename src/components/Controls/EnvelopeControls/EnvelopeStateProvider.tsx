import React, { PropsWithChildren, useReducer } from 'react';
import initialState, { EnvelopeState } from './store/initialState';
import { ActionTypes } from './store/actions';
import reducer from './store/reducer';

interface EnvelopeStateProviderValue {
  state: EnvelopeState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const envelopeStateContext =
  React.createContext<EnvelopeStateProviderValue>(undefined);

export const EnvelopeStateProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <envelopeStateContext.Provider value={{ state, dispatch }}>
      {children}
    </envelopeStateContext.Provider>
  );
};
