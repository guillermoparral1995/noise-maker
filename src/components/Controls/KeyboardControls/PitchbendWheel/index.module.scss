#pitchbend_wheel_container {
    margin: 20px;
    text-align: center;
}

#pitchbend_wheel {
    &.desktop {
        transform: rotate(270deg);
    }
    appearance: none;
    height: 3em;
    width: 12em;
    box-shadow: 1px 1px 5px mediumpurple;
}

#pitchbend_wheel::-webkit-slider-thumb {
    -webkit-appearance: none;
    width: 10px;
    height: 15px;
    background: mediumpurple;
    border-radius: 4px;
    cursor: -moz-grab;
    cursor: -webkit-grab;
    cursor:move;
}