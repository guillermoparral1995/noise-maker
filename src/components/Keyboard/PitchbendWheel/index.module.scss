#pitchbend_wheel {
    transform: rotate(270deg);
    appearance: none;
    height: 3em;
    width: 12em;
    margin: -10px;
    box-shadow: 1px 1px 5px mediumpurple;
}

#pitchbend_wheel::-webkit-slider-thumb {
    -webkit-appearance: none;
    width: 10px;
    height: 15px;
    background: mediumpurple;
    border-radius: 4px;
    cursor: -moz-grab;
    cursor: -webkit-grab;
    cursor:move;
}