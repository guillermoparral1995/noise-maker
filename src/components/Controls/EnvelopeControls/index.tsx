import useAddMidiListeners from '@hooks/useAddMidiListeners';
import { Knobs } from '@types';
import React, { useContext } from 'react';
import ControlsRow from '../../shared/ControlsRow';
import Knob from '../../shared/Knob';
import { envelopeStateContext } from './EnvelopeStateProvider';

const EnvelopeControls = () => {
  const { state, dispatch } = useContext(envelopeStateContext);
  useAddMidiListeners(
    [Knobs.ATTACK, Knobs.DECAY, Knobs.SUSTAIN, Knobs.RELEASE],
    dispatch,
  );

  return (
    <>
      <h3>ADSR Envelope</h3>
      <ControlsRow>
        <Knob id={Knobs.ATTACK} value={state.attack} dispatch={dispatch}></Knob>
        <Knob id={Knobs.DECAY} value={state.decay} dispatch={dispatch}></Knob>
        <Knob
          id={Knobs.SUSTAIN}
          value={state.sustain}
          dispatch={dispatch}
        ></Knob>
        <Knob
          id={Knobs.RELEASE}
          value={state.release}
          dispatch={dispatch}
        ></Knob>
      </ControlsRow>
    </>
  );
};

export default EnvelopeControls;
