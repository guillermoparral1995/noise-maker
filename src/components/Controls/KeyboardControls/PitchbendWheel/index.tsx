import { knobsValues } from '@constants/knobsValues';
import { useBreakpoints } from '@hooks/useBreakpoints';
import { midiContext } from '@providers/MIDIProvider';
import { Knobs } from '@types';
import { round } from 'lodash';
import React, { BaseSyntheticEvent, useContext, useEffect } from 'react';
import { envelopeStateContext } from '../../EnvelopeControls/EnvelopeStateProvider';
import { updatePitchbend } from '../../EnvelopeControls/store/actions';
import styles from './index.module.scss';

export const PitchbendWheel = () => {
  const { selectedInput: midiInput } = useContext(midiContext);
  const { state, dispatch } = useContext(envelopeStateContext);
  const { isDesktop } = useBreakpoints();

  useEffect(() => {
    if (midiInput) {
      midiInput.addListener('pitchbend', (e) => {
        const multiplier =
          (knobsValues[Knobs.PITCHBEND].max -
            knobsValues[Knobs.PITCHBEND].min) /
          2;
        const value = round((e.value as number) * multiplier, 2);
        dispatch(updatePitchbend(value));
      });
    }
  }, [midiInput]);

  const handleChange = (e: BaseSyntheticEvent) => {
    const pitchbendValue = round(e.target.value as number, 2);
    dispatch(updatePitchbend(pitchbendValue));
  };

  const handleDragEnd = () => {
    dispatch(updatePitchbend(0));
  };

  return (
    <div id={styles.pitchbend_wheel_container}>
      <input
        id={styles.pitchbend_wheel}
        className={isDesktop ? styles.desktop : ''}
        data-testid="pitchbend_wheel"
        type="range"
        min={knobsValues[Knobs.PITCHBEND].min}
        max={knobsValues[Knobs.PITCHBEND].max}
        step={knobsValues[Knobs.PITCHBEND].step}
        value={state.pitchbend}
        onChange={handleChange}
        onMouseUp={handleDragEnd}
      ></input>
    </div>
  );
};

export default PitchbendWheel;
