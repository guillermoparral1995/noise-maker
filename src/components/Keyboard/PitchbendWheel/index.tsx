import { round } from 'lodash';
import React, { BaseSyntheticEvent, useContext, useEffect } from 'react';
import { knobsValues } from '../../../constants/knobsValues';
import { midiContext } from '../../../providers/MIDIProvider';
import { Knobs } from '../../../types';
import { envelopeStateContext } from '../../Controls/EnvelopeControls/EnvelopeStateProvider';
import { updatePitchbend } from '../../Controls/EnvelopeControls/store/actions';
import styles from './index.module.scss';

export const PitchbendWheel = () => {
  const { selectedInput: midiInput } = useContext(midiContext);
  const { state, dispatch } = useContext(envelopeStateContext);

  useEffect(() => {
    if (midiInput) {
      midiInput.addListener('pitchbend', (e) => {
        const multiplier =
          (knobsValues[Knobs.PITCHBEND].max -
            knobsValues[Knobs.PITCHBEND].min) /
          2;
        const value = round((e.value as number) * multiplier, 2);
        dispatch(updatePitchbend(value));
      });
    }
  }, [midiInput]);

  const handleChange = (e: BaseSyntheticEvent) => {
    const pitchbendValue = round(e.target.value as number, 2);
    dispatch(updatePitchbend(pitchbendValue));
  };

  const handleDragEnd = () => {
    dispatch(updatePitchbend(0));
  };

  return (
    <input
      id={styles.pitchbend_wheel}
      type="range"
      min={knobsValues[Knobs.PITCHBEND].min}
      max={knobsValues[Knobs.PITCHBEND].max}
      step={knobsValues[Knobs.PITCHBEND].step}
      value={state.pitchbend}
      onChange={handleChange}
      onMouseUp={handleDragEnd}
    ></input>
  );
};

export default PitchbendWheel;
