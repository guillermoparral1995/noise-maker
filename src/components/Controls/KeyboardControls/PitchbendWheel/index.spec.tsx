import { withMockedMIDIInput, withMockedMIDINoInput } from '@mocks';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Actions } from '@types';
import React from 'react';
import PitchbendWheel from '.';
import { EnvelopeStateProvider } from '../../EnvelopeControls/EnvelopeStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

describe('Pitchbend wheel', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render correctly', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <EnvelopeStateProvider>
          <PitchbendWheel></PitchbendWheel>
        </EnvelopeStateProvider>,
      ),
    );

    const pitchbendWheel = await findByTestId('pitchbend_wheel');
    expect(pitchbendWheel).toMatchSnapshot();
  });

  it('should dispatch event when changing input', async () => {
    const mockDispatch = jest.fn();
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <EnvelopeStateProvider __mockDispatch={mockDispatch}>
          <PitchbendWheel></PitchbendWheel>
        </EnvelopeStateProvider>,
      ),
    );

    const pitchbendWheel = await findByTestId('pitchbend_wheel');
    fireEvent.change(pitchbendWheel, { target: { value: 100 } });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: 100,
      type: Actions.UPDATE_PITCHBEND,
    });
  });

  it('should dispatch event when dragging ends', async () => {
    const mockDispatch = jest.fn();
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <EnvelopeStateProvider __mockDispatch={mockDispatch}>
          <PitchbendWheel></PitchbendWheel>
        </EnvelopeStateProvider>,
      ),
    );

    const pitchbendWheel = await findByTestId('pitchbend_wheel');
    fireEvent.mouseDown(pitchbendWheel);
    fireEvent.mouseUp(pitchbendWheel);

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: 0,
      type: Actions.UPDATE_PITCHBEND,
    });
  });

  it('should dispatch event when using MIDI input', async () => {
    const mockDispatch = jest.fn();
    let mockEventName: string;
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        mockEventName = eventName;
        listener({ value: 0.5 });
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <EnvelopeStateProvider __mockDispatch={mockDispatch}>
          <PitchbendWheel></PitchbendWheel>
        </EnvelopeStateProvider>,
      ),
    );

    await findByTestId('pitchbend_wheel');

    expect(mockAddListener).toHaveBeenCalled();
    expect(mockEventName).toEqual('pitchbend');
    expect(mockDispatch).toHaveBeenCalledWith({
      payload: 150,
      type: Actions.UPDATE_PITCHBEND,
    });
  });
});
