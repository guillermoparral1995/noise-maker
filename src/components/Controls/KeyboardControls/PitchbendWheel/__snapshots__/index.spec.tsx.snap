// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pitchbend wheel should render correctly 1`] = `
<input
  class="desktop"
  data-testid="pitchbend_wheel"
  id="pitchbend_wheel"
  max="300"
  min="-300"
  step="1"
  type="range"
  value="0"
/>
`;
